internal/sim/api_test.go

```diff
@@ -0,0 +1,204 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sim
+
+import (
+	"context"
+	"io"
+	"math"
+	"math/rand"
+	"reflect"
+	"strings"
+	"testing"
+
+	"github.com/ServiceWeaver/weaver/internal/reflection"
+)
+
+// TODO(mwhittaker): Move these into the sim package.
+type integers struct{}
+type positives struct{}
+type float64s struct{}
+
+var _ Generator[int] = integers{}
+var _ Generator[int] = positives{}
+var _ Generator[float64] = float64s{}
+
+func (integers) Generate(r *rand.Rand) int     { return r.Int() }
+func (positives) Generate(r *rand.Rand) int    { return 1 + r.Intn(math.MaxInt) }
+func (float64s) Generate(r *rand.Rand) float64 { return r.Float64() }
+
+type divModWorkload struct{}
+
+func (d *divModWorkload) Init(r Registrar) error {
+	r.RegisterGenerators("DivMod", integers{}, positives{})
+	r.RegisterGenerators("Div", integers{}, positives{})
+	r.RegisterGenerators("Mod", integers{}, positives{})
+	return nil
+}
+
+func (d *divModWorkload) DivMod(context.Context, int, int) error { return nil }
```

Comment:
Have you considered passing Generators directly here, in the method signature?

I can see it being weird because these methods will be called with generated values, not with generators. But maybe it can be done this way, with global type registration:

```
type positive int
type positives struct{}
func (positives) Generate(r *rand.Rand) positive { return 1 + r.Intn(math.MaxInt)}
type integers struct{}
func (integers) Generate(r *rand.Rand) int { return r.Int() }

func init() {
  RegisterGenerator(integers{})
  RegisterGenerator(positives{})
}

func (d *divModWorkload) DivMod(context.Context, int, positive) error { return nil }
```

Reply:
Good question. We definitely explored the idea of methods receiving generators. I disliked the idea for a couple of reasons. Primarily, I placed a high value on method signatures being as simple and as direct as possible.

- In the body of your `DivMod`, for example, the user would have to cast `positive` to `int`. This is a bit roundabout and also error-prone as they might cast to the wrong type (e.g., `int32`). 
- We could work around this by having `positive` (or `positives`) have a `Get` method or something similar that returns an `int`, but that complicates things further. 
- In your `DivMod` example, I also feel it's a bit discordant that the first argument gets to be a plain `int` while `positive` does not. If you later change how you want to generate the first argument, you'd have to change its type and also update the body of `DivMod` to perform casting (or calling a `Get` method).

Overall, I felt our approach requires more typing but is also more straightforward with fewer subtleties. My plan is to stick with registering generators in `Init` for now, get the simulator working on some examples, and reconsider the API once I have a bit more experience with it. How does that sound?

Reply:
Good point about type casting. I'm okay with us moving forward with what you have.

Whatever we do, I like that both your API and my proposal have the property that it is clear these methods signatures are primitive values, and not some types like `Gen[int]`. The conceptual model for the user should be that once the method is called, the values have already been generated, and nothing conveys that better than the signature consisting of primitive types.